README.rst
================
Yet Another Cron
================


A modern Cron replacement that is Docker-friendly


* Free software: MIT license

Features
--------

* "Crontab" is in YAML format;
* Builtin sending of Sentry and Mail outputs when cron jobs fail;
* Flexible configuration: you decide how to determine if a cron job fails or not;
* Designed for running in Docker, Kubernetes, or 12 factor environments:

  * Runs in the foreground;
  * Logs everything to stdout/stderr [1]_;

* Option to automatically retry failing cron jobs, with exponential backoff;
* Optional HTTP REST API, to fetch status and start jobs on demand;
* Arbitrary timezone support;


.. [1] Whereas vixie cron only logs to syslog, requiring a syslog daemon to be running in the background or else you don't get logs!

Status
--------------

The project is in beta stage: essential features are complete, and the focus is
finding and fixing bugs before the first stable release.

Installation
------------

Install using pip
+++++++++++++++++

yacron requires Python >= 3.6 (for systems with older Python, use the binary instead).  It is advisable to install it in a Python
virtual environment, for example:

.. code-block:: shell

    python3 -m venv yacronenv
    . yacronenv/bin/activate
    pip install yacron

Install using pipx
++++++++++++++++++

pipx_ automates creating a virtualenv and installing a python program in the
newly created virtualenv.  It is as simple as:

.. code-block:: shell

    pipx install yacron

.. _pipx: https://github.com/pipxproject/pipx

Install using binary
++++++++++++++++++++

Alternatively, a self-contained binary can be downloaded
from github: https://github.com/gjcarneiro/yacron/releases. This binary should
work on any Linux 64-bit system post glibc 2.23 (e.g. Ubuntu:16.04).  Python is not required on the target system (it is embedded in the executable).

Usage
-----

Configuration is in YAML format.  To start yacron, give it a configuration file
or directory path as the ``-c`` argument.  For example::

    yacron -c /tmp/my-crontab.yaml

This starts yacron (always in the foreground!), reading
``/tmp/my-crontab.yaml`` as configuration file.  If the path is a directory,
any ``*.yaml`` or ``*.yml`` files inside this directory are taken as
configuration files.

Configuration basics
++++++++++++++++++++

This configuration runs a command every 5 minutes:

.. code-block:: yaml

    jobs:
      - name: test-01
        command: echo "foobar"
        shell: /bin/bash
        schedule: "*/5 * * * *"

The command can be a string or a list of strings.  If command is a string,
yacron runs it through a shell, which is ``/bin/bash`` in the above example, but
is ``/bin/sh`` by default.

If the command is a list of strings, the command is executed directly, without a
shell.  The ARGV of the command to execute is extracted directly from the
configuration:

.. code-block:: yaml

    jobs:
      - name: test-01
        command:
          - echo
          - foobar
        schedule: "*/5 * * * *"


The `schedule` option can be a string in a crontab format specified by https://github.com/josiahcarlson/parse-crontab (this module is used by yacron).
Additionally @reboot can be included , which will only run the job when yacron is initially
executed. Further `schedule` can be an object with properties.  The following configuration
runs a command every 5 minutes, but only on the specific date 2017-07-19, and
doesn't run it in any other date:

.. code-block:: yaml

    jobs:
      - name: test-01
        command: echo "foobar"
        schedule:
          minute: "*/5"
          dayOfMonth: 19
          month: 7
          year: 2017
          dayOfWeek: "*"

Important: by default all time is interpreted to be in UTC, but you can
request to use local time instead.  For instance, the cron job below runs
every day at 19h27 *local time* because of the ``utc: false`` option:

.. code-block:: yaml

  jobs:
    - name: test-01
      command: echo "hello"
      schedule: "27 19 * * *"
      utc: false
      captureStdout: true

Since Yacron version 0.11, you can also request that the schedule be
interpreted in an arbitrary timezone, using the ``timezone`` attribute:

.. code-block:: yaml

  jobs:
    - name: test-01
      command: echo "hello"
      schedule: "27 19 * * *"
      timezone: America/Los_Angeles
      captureStdout: true


You can ask for environment variables to be defined for command execution:

.. code-block:: yaml

    jobs:
      - name: test-01
        command: echo "foobar"
        shell: /bin/bash
        schedule: "*/5 * * * *"
        environment:
          - key: PATH
            value: /bin:/usr/bin

You can also provide an environment file to define environments for command execution:

.. code-block:: yaml

    jobs:
      - name: test-01
        command: echo "foobar"
        shell: /bin/bash
        schedule: "*/5 * * * *"
        env_file: .env

The env file must be a list of ``KEY=VALUE`` pairs. Empty lines and lines starting with ``#`` will be ignored.

Variables declared in the ``environment`` option will override those found in the ``env_file``.


Specifying defaults
+++++++++++++++++++


There can be a special ``defaults`` section in the config.  Any attributes
defined in this section provide default values for cron jobs to inherit.
Although cron jobs can still override the defaults, as needed:

.. code-block:: yaml

    defaults:
        environment:
          - key: PATH
            value: /bin:/usr/bin
        shell: /bin/bash
        utc: false
    jobs:
      - name: test-01
        command: echo "foobar"  # runs with /bin/bash as shell
        schedule: "*/5 * * * *"
      - name: test-02  # runs with /bin/sh as shell
        command: echo "zbr"
        shell: /bin/sh
        schedule: "*/5 * * * *"

Note: if the configuration option is a directory and there are multiple configuration files in that directory, then the ``defaults`` section in each configuration file provides default options only for cron jobs inside that same file; the defaults have no effect beyond any individual YAML file.

Reporting
+++++++++

Yacron has builtin support for reporting jobs failure (more on that below) by
email, Sentry and shell command (additional reporting methods might be added in the future):

.. code-block:: yaml

  - name: test-01
    command: |
      echo "hello" 1>&2
      sleep 1
      exit 10
    schedule:
      minute: "*/2"
    captureStderr: true
    onFailure:
      report:
        sentry:
          dsn:
            value: example
            # Alternatively:
            # fromFile: /etc/secrets/my-secret-dsn
            # fromEnvVar: SENTRY_DSN
          fingerprint:  # optional, since yacron 0.6
            - yacron
            - "{{ environment.HOSTNAME }}"
            - "{{ name }}"
          extra:
            foo: bar
            zbr: 123
          level: warning
          environment: production
        mail:
          from: example@foo.com
          to: example@bar.com
          smtpHost: 127.0.0.1
          # optional fields:
          username: "username1"  # set username and password to enable login
          password:
            value: example
            # Alternatively:
            # fromFile: /etc/secrets/my-secret-password
            # fromEnvVar: MAIL_PASSWORD
          tls: false  # set to true to enable TLS
          starttls: false  # set to true to enable StartTLS
        shell:
          shell: /bin/bash
          command: ...

Here, the ``onFailure`` object indicates that what to do when a job failure
is detected.  In this case we ask for it to be reported both to sentry and by
sending an email.

The ``captureStderr: true`` part instructs yacron to capture output from the the
program's `standard error`, so that it can be included in the report.  We could
also turn on `standard output` capturing via the ``captureStdout: true`` option.
By default, yacron captures only standard error.  If a cron job's standard error
or standard output capturing is not enabled, these streams will simply write to
the same standard output and standard error as yacron itself.

Both `stdout` and `stderr` stream lines are by default prefixed with
``[{job_name} {stream_name}]``, i.e. ``[test-01 stdout]``, if for any reason you
need to change this, provide the option ``streamPrefix`` (new in version 0.16)
with your own custom string.

.. code-block:: yaml

  - name: test-01
    command: echo "hello world"
    schedule:
      minute: "*/2"
    captureStdout: true
    streamPrefix: "[{job_name} job]"

In some cases, for instance when you're logging JSON objects you might want to
completely get rid of the prefix altogether:

.. code-block:: yaml

  - name: test-01
    command: echo "hello world"
    schedule:
      minute: "*/2"
    captureStdout: true
    streamPrefix: ""

It is possible also to report job success, as well as failure, via the
``onSuccess`` option.

.. code-block:: yaml

  - name: test-01
    command: echo "hello world"
    schedule:
      minute: "*/2"
    captureStdout: true
    onSuccess:
      report:
        mail:
          from: example@foo.com
          to: example@bar.com
          smtpHost: 127.0.0.1

Since yacron 0.5, it is possible to customise the format of the report. For
``mail`` reporting, the option ``subject`` indicates what is the subject of the
email, while ``body`` formats the email body.  For Sentry reporting, there is
only ``body``.  In all cases, the values of those options are strings that are
processed by the jinja2_ templating engine.  The following variables are
available in templating:

* name(str): name of the cron job
* success(bool): whether or not the cron job succeeded
* stdout(str): standard output of the process
* stderr(str): standard error of the process
* exit_code(int): process exit code
* command(str): cron job command
* shell(str): cron job shell
* environment(dict): subprocess environment variables

.. _jinja2: http://jinja.pocoo.org/

Example:

.. code-block:: yaml

  - name: test-01
    command: |
      echo "hello" 1>&2
      sleep 1
      exit 10
    schedule:
      minute: "*/2"
    captureStderr: true
    onFailure:
      report:
        mail:
          from: example@foo.com
          to: example@bar.com
          smtpHost: 127.0.0.1
          subject: Cron job '{{name}}' {% if success %}completed{% else %}failed{% endif %}
          body: |
            {{stderr}}
            (exit code: {{exit_code}})


The shell reporter (since yacron 0.13) executes a user given shell command in
the specified shell. It passes all environment variables from the python
executable and specifies some additional ones to inform about the state of the
job:

* YACRON_FAIL_REASON (str)
* YACRON_FAILED ("1" or "0")
* YACRON_JOB_NAME (str)
* YACRON_JOB_COMMAND (str)
* YACRON_JOB_SCHEDULE (str)
* YACRON_RETCODE (str)
* YACRON_STDERR (str)
* YACRON_STDOUT (str)

A simple example configuration:

.. code-block:: yaml

  - name: test-01
    command: echo "foobar" && exit 123
    shell: /bin/bash
    schedule: "* * * * *"
    onFailure:
      report:
        shell:
          shell: /bin/bash
          command: echo "Error code $YACRON_RETCODE"

Since yacron 0.15, it is possible to send emails formatted as html, by  adding
the ``html: true`` property.  For example, here the standard output of a shell
command is captured and interpreted as html and placed in the email message\:

.. code-block:: yaml

  - name: test-01
    command: echo "hello <b>world</b>"
    schedule: "@reboot"
    captureStdout: true
    onSuccess:
      report:
        mail:
          from: example@foo.com
          to: example@bar.com, zzz@sleep.com
          html: true
          smtpHost: 127.0.0.1
          smtpPort: 1025
          subject: This is a cron job with html body


Metrics
+++++++++

Yacron has builtin support for writing job metrics to Statsd_:

.. _Statsd: https://github.com/etsy/statsd

.. code-block:: yaml

    jobs:
      - name: test01
        command: echo "hello"
        schedule: "* * * * *"
        statsd:
          host: my-statsd.example.com
          port: 8125
          prefix: my.cron.jobs.prefix.test01

With this config Yacron will write the following metrics over UDP
to the Statsd listening on ``my-statsd.example.com:8125``:

.. code-block::

  my.cron.jobs.prefix.test01.start:1|g  # this one is sent when the job starts
  my.cron.jobs.prefix.test01.stop:1|g   # the rest are sent when the job stops
  my.cron.jobs.prefix.test01.success:1|g
  my.cron.jobs.prefix.test01.duration:3|ms|@0.1


Handling failure
++++++++++++++++

By default, yacron considers that a job has `failed` if either the process
returns a non-zero code or if it generates output to `standard error` (and
standard error capturing is enabled, of course).

You can instruct yacron how to determine if a job has failed or not via the
``failsWhen`` option:

.. code-block:: yaml

  failsWhen:
    producesStdout: false
    producesStderr: true
    nonzeroReturn: true
    always: false

producesStdout
    If true, any captured standard output causes yacron to consider the job
    as failed.  This is false by default.

producesStderr
    If true, any captured standard error causes yacron to consider the job
    as failed.  This is true by default.

nonzeroReturn
    If true, if the job process returns a code other than zero causes yacron
    to consider the job as failed.  This is true by default.

always
    If true, if the job process exits that causes yacron to consider the job as
    failed.  This is false by default.

It is possible to instruct yacron to retry failing cron jobs by adding a
``retry`` option inside ``onFailure``:

.. code-block:: yaml

  - name: test-01
    command: |
      echo "hello" 1>&2
      sleep 1
      exit 10
    schedule:
      minute: "*/10"
    captureStderr: true
    onFailure:
      report:
        mail:
          from: example@foo.com
          to: example@bar.com
          smtpHost: 127.0.0.1
      retry:
        maximumRetries: 10
        initialDelay: 1
        maximumDelay: 30
        backoffMultiplier: 2

The above settings tell yacron to retry the job up to 10 times, with the delay
between retries defined by an exponential backoff process: initially 1 second,
doubling for every retry up to a maximum of 30 seconds. A value of -1 for
maximumRetries will mean yacron will keep retrying forever, this is mostly

HISTORY.rst
=======
History
=======

0.19.0 (2023-03-11)
-------------------

* Add ability to configure yacron's own logging (#81 #82 #83, gjcarneiro, bdamian)
* Add config value for SMTP(validate_certs=False) (David Batley)

0.18.0 (2023-01-01)
-------------------

* fixes "Job is always executed immediately on yacron start" (#67)
* add an `enabled` option in jobs (#73)
* give a better error message when no configuration file is provided or exists (#72)

0.17.0 (2022-06-26)
-------------------

* Support Additional Shell Report Vars (RJ Garcia)
* Shell reporter: handle long lines truncatation (Hannes Hergeth)
* exe: undo pyinstaller LD_LIBRARY_PATH changes in subprocesses (#68, Gustavo Carneiro)


0.16.0 (2021-12-05)
-------------------

* make the capture max line length configurable and change the default
  from 64K to 16M (#56)
* Add config option to change prefix of subprocess stream lines (#58, eelkeh)


0.15.1 (2021-11-19)
-------------------

* Fix a bug in the --validate option (#57, Leonid Repin)


0.15.0 (2021-11-10)
-------------------

* Allow emails to be html formatted
* Fix an error when reading cmd output with huge lines (#56)


0.14.0 (2021-10-04)
-------------------

* Sentry: increase the size of messages before getting truncated #54
* Sentry: allow specifying the environment option #53
* Minor fixes


0.13.1 (2021-08-10)
-------------------

* unicode fixes for the exe binary version

0.13.0 (2021-06-28)

* Add ability for one config file to include another one #38
* Add shell command reporting ability (Hannes Hergeth, #50)


0.12.2 (2021-05-31)
-------------------

* constrain ruamel.yaml to version 0.17.4 or below, later versions are buggy


0.12.1 (2021-05-30)
-------------------

* blacklist ruamel.yaml version 0.17.5 in requirements #47


0.12.0 (2021-04-22)
-------------------

* web: don't crash when receiving a web request without Accept header (#45)
* add env_file configuration option (Alessandro Romani, #43)
* email: add missing Date header (#39)


0.11.2 (2020-11-29)
-------------------

* Add back a self contained binary, this time based on PyInstaller

0.11.1 (2020-07-29)
-------------------

* Fix email reporting when multiple recipients given


0.11.0 (2020-07-20)
-------------------

* reporting: add a failure reason line at the top of sentry/email (#36)
* mail: new tls, startls, username, and password options (#21)
* allow jobs to run as a different user (#18)
* Support timezone schedule (#26)


0.10.1 (2020-06-02)
-------------------

* Minor bugfixes


0.10.0 (2019-11-03)
-------------------

* HTTP remote interface, allowing to get job status and start jobs on demand
* Simple Linux binary including all dependencies (built using PyOxidizer)

0.10.0b2 (2019-10-26)
---------------------

* Build Linux binary inside Docker Ubuntu 16.04, so that it is compatible with
  older glibc systems

0.10.0b1 (2019-10-13)
---------------------
* Build a standalone Linux binary, using PyOxidizer
* Switch from raven to sentry-sdk

0.9.0 (2019-04-03)
------------------
* Added an option to just check if the yaml file is valid without running the scheduler.
* Fix missing `body` in the schema for sentry config


0.8.1 (2018-10-16)
------------------
* Fix a bug handling ``@reboot`` in schedule (#22)

0.8.0 (2018-05-14)
------------------
* Sentry: add new ``extra`` and ``level`` options.


0.7.0 (2018-03-21)
------------------

* Added the ``utc`` option and document that times are utc by default (#17);
* If an email body is empty, skip sending it;
* Added docker and k8s example.


0.6.0 (2017-11-24)
------------------
* Add custom Sentry fingerprint support
* Ability to send job metrics to statsd (thanks bofm)
* ``always`` flag to consider any cron job that exits to be failed
  (thanks evanjardineskinner)
* `maximumRetries` can now be ``-1`` to never stop retrying (evanjardineskinner)
* ``schedule`` can be the string ``@reboot`` to always run that cron job on startup
  (evanjardineskinner)
* ``saveLimit`` can be set to zero (evanjardineskinner)

0.5.0
------------------
* Templating support for reports
* Remove deprecated smtp_host/smtp_port

0.4.3 (2017-09-13)
------------------
* Bug fixes

0.4.2 (2017-09-07)
------------------
* Bug fixes

0.4.1 (2017-08-03)
------------------

* More polished handling of configuration errors;
* Unit tests;
* Bug fixes.

0.4.0 (2017-07-24)
------------------

* New option ``executionTimeout``, to terminate jobs that get stuck;
* If a job doesn't terminate gracefully kill it.  New option ``killTimeout``
  controls how much time to wait for graceful termination before killing it;
* Switch parsing to strictyaml, for more user friendly parsing validation error
  messages.


yacron/config.py
from dataclasses import dataclass
from pwd import getpwnam
from grp import getgrnam
import logging
import os.path
from typing import Union  # noqa
from typing import List, Optional, Any, Dict, NewType
import datetime
import pytz

import strictyaml
from strictyaml import Optional as Opt, EmptyDict
from strictyaml import Any as YamlAny
from strictyaml import (
    Bool,
    EmptyNone,
    Enum,
    Float,
    Int,
    Map,
    Seq,
    Str,
    MapPattern,
)
from strictyaml.ruamel.error import YAMLError

from crontab import CronTab

logger = logging.getLogger("yacron.config")
WebConfig = NewType("WebConfig", Dict[str, Any])
JobDefaults = NewType("JobDefaults", Dict[str, Any])
LoggingConfig = NewType("LoggingConfig", Dict[str, Any])


class ConfigError(Exception):
    pass


DEFAULT_BODY_TEMPLATE = """
{% if fail_reason -%}
(job failed because {{fail_reason}})
{% endif %}
{% if stdout and stderr -%}
STDOUT:
---
{{stdout}}
---
STDERR:
{{stderr}}
{% elif stdout -%}
{{stdout}}
{% elif stderr -%}
{{stderr}}
{% else -%}
(no output was captured)
{% endif %}
"""

DEFAULT_SUBJECT_TEMPLATE = (
    "Cron job '{{name}}' {% if success %}completed"
    "{% else %}failed{% endif %}"
)

_REPORT_DEFAULTS = {
    "sentry": {
        "dsn": {"value": None, "fromFile": None, "fromEnvVar": None},
        "body": DEFAULT_SUBJECT_TEMPLATE + "\n" + DEFAULT_BODY_TEMPLATE,
        "fingerprint": ["yacron", "{{ environment.HOSTNAME }}", "{{ name }}"],
        "environment": None,
        "maxStringLength": 8192,
    },
    "mail": {
        "from": None,
        "to": None,
        "smtpHost": None,
        "smtpPort": 25,
        "tls": False,
        "starttls": False,
        "validate_certs": False,
        "html": False,
        "subject": DEFAULT_SUBJECT_TEMPLATE,
        "body": DEFAULT_BODY_TEMPLATE,
        "username": None,
        "password": {"value": None, "fromFile": None, "fromEnvVar": None},
    },
    "shell": {
        "shell": "/bin/sh",
        "command": None,
    },
}


DEFAULT_CONFIG = {
    "shell": "/bin/sh",
    "concurrencyPolicy": "Allow",
    "captureStderr": True,
    "captureStdout": False,
    "saveLimit": 4096,
    "maxLineLength": 16 * 1024 * 1024,
    "utc": True,
    "timezone": None,
    "failsWhen": {
        "producesStdout": False,
        "producesStderr": True,
        "nonzeroReturn": True,
        "always": False,
    },
    "onFailure": {
        "retry": {
            "maximumRetries": 0,
            "initialDelay": 1,
            "maximumDelay": 300,
            "backoffMultiplier": 2,
        },
        "report": _REPORT_DEFAULTS,
    },
    "onPermanentFailure": {"report": _REPORT_DEFAULTS},
    "onSuccess": {"report": _REPORT_DEFAULTS},
    "environment": [],
    "env_file": None,
    "executionTimeout": None,
    "killTimeout": 30,
    "statsd": None,
    "streamPrefix": "[{job_name} {stream_name}] ",
    "enabled": True,
}


_report_schema = Map(
    {
        Opt("sentry"): Map(
            {
                Opt("dsn"): Map(
                    {
                        Opt("value"): EmptyNone() | Str(),
                        Opt("fromFile"): EmptyNone() | Str(),
                        Opt("fromEnvVar"): EmptyNone() | Str(),
                    }
                ),
                Opt("fingerprint"): Seq(Str()),
                Opt("level"): Str(),
                Opt("extra"): MapPattern(Str(), Str() | Int() | Bool()),
                Opt("body"): Str(),
                Opt("environment"): Str(),
                Opt("maxStringLength"): Int(),
            }
        ),
        Opt("mail"): Map(
            {
                "from": EmptyNone() | Str(),
                "to": EmptyNone() | Str(),
                Opt("smtpHost"): Str(),
                Opt("smtpPort"): Int(),
                Opt("subject"): Str(),
                Opt("body"): Str(),
                Opt("username"): Str(),
                Opt("password"): Map(
                    {
                        Opt("value"): EmptyNone() | Str(),
                        Opt("fromFile"): EmptyNone() | Str(),
                        Opt("fromEnvVar"): EmptyNone() | Str(),
                    }
                ),
                Opt("tls"): Bool(),
                Opt("starttls"): Bool(),
                Opt("validate_certs"): Bool(),
                Opt("html"): Bool(),
            }
        ),
        Opt("shell"): Map(
            {
                Opt("shell"): Str(),
                "command": Str() | Seq(Str()),
            }
        ),
    }
)

_job_defaults_common = {
    Opt("shell"): Str(),
    Opt("concurrencyPolicy"): Enum(["Allow", "Forbid", "Replace"]),
    Opt("captureStderr"): Bool(),
    Opt("captureStdout"): Bool(),
    Opt("saveLimit"): Int(),
    Opt("maxLineLength"): Int(),
    Opt("utc"): Bool(),
    Opt("timezone"): Str(),
    Opt("failsWhen"): Map(
        {
            "producesStdout": Bool(),
            Opt("producesStderr"): Bool(),
            Opt("nonzeroReturn"): Bool(),
            Opt("always"): Bool(),
        }
    ),
    Opt("onFailure"): Map(
        {
            Opt("retry"): Map(
                {
                    "maximumRetries": Int(),
                    "initialDelay": Float(),
                    "maximumDelay": Float(),
                    "backoffMultiplier": Float(),
                }
            ),
            Opt("report"): _report_schema,
        }
    ),
    Opt("onPermanentFailure"): Map({Opt("report"): _report_schema}),
    Opt("onSuccess"): Map({Opt("report"): _report_schema}),
    Opt("environment"): Seq(Map({"key": Str(), "value": Str()})),
    Opt("env_file"): Str(),
    Opt("executionTimeout"): Float(),
    Opt("killTimeout"): Float(),
    Opt("statsd"): Map({"prefix": Str(), "host": Str(), "port": Int()}),
    Opt("user"): Str() | Int(),
    Opt("group"): Str() | Int(),
    Opt("streamPrefix"): Str(),
    Opt("enabled"): Bool(),
}

_job_schema_dict = dict(_job_defaults_common)
_job_schema_dict.update(
    {
        "name": Str(),
        "command": Str() | Seq(Str()),
        "schedule": Str()
        | Map(
            {
                Opt("minute"): Str(),
                Opt("hour"): Str(),
                Opt("dayOfMonth"): Str(),
                Opt("month"): Str(),
                Opt("year"): Str(),
                Opt("dayOfWeek"): Str(),
            }
        ),
    }
)

CONFIG_SCHEMA = EmptyDict() | Map(
    {
        Opt("defaults"): Map(_job_defaults_common),
        Opt("jobs"): Seq(Map(_job_schema_dict)),
        Opt("web"): Map({"listen": Seq(Str())}),
        Opt("include"): Seq(Str()),
        Opt("logging"): Map(
            {
                "version": Int(),
                Opt("incremental"): Bool(),
                Opt("disable_existing_loggers"): Bool(),
                Opt("formatters"): YamlAny(),
                Opt("filters"): YamlAny(),
                Opt("handlers"): YamlAny(),
                Opt("loggers"): YamlAny(),
                Opt("root"): YamlAny(),
            }
        ),
    }
)


# Slightly modified version of https://stackoverflow.com/a/7205672/2211825
def mergedicts(dict1, dict2):
    for k in set(dict1.keys()).union(dict2.keys()):
        if k in dict1 and k in dict2:
            v1 = dict1[k]
            v2 = dict2[k]
            if isinstance(v1, dict) and isinstance(v2, dict):
                yield (k, dict(mergedicts(v1, v2)))
            elif isinstance(v1, dict) and v2 is None:  # modification
                yield (k, dict(mergedicts(v1, {})))
            elif isinstance(v1, list) and isinstance(v2, list):  # merge lists
                yield (k, v1 + v2)
            else:
                yield (k, v2)
        elif k in dict1:
            yield (k, dict1[k])
        else:
            yield (k, dict2[k])


class JobConfig:
    def __init__(self, config: dict) -> None:
        self.name = config["name"]  # type: str
        self.command = config["command"]  # type: Union[str, List[str]]
        self.schedule_unparsed = config.pop("schedule")
        if isinstance(self.schedule_unparsed, str):
            if self.schedule_unparsed in {"@reboot"}:
                self.schedule = (
                    self.schedule_unparsed
                )  # type: Union[CronTab, str]
            else:
                self.schedule = CronTab(self.schedule_unparsed)
        elif isinstance(self.schedule_unparsed, dict):
            minute = self.schedule_unparsed.get("minute", "*")
            hour = self.schedule_unparsed.get("hour", "*")
            day = self.schedule_unparsed.get("dayOfMonth", "*")
            month = self.schedule_unparsed.get("month", "*")
            dow = self.schedule_unparsed.get("dayOfWeek", "*")
            tab = "{} {} {} {} {}".format(minute, hour, day, month, dow)
            logger.debug("Converted schedule to %r", tab)
            self.schedule = CronTab(tab)
        else:
            raise ConfigError("invalid schedule: %r", self.schedule_unparsed)
        self.shell = config.pop("shell")
        self.concurrencyPolicy = config.pop("concurrencyPolicy")
        self.captureStderr = config.pop("captureStderr")
        self.captureStdout = config.pop("captureStdout")
        self.streamPrefix = config.pop("streamPrefix")
        self.saveLimit = config.pop("saveLimit")
        self.maxLineLength = config.pop("maxLineLength")
        self.utc = config.pop("utc")
        self.enabled: bool = config.pop("enabled")
        self.timezone = None  # type: Optional[datetime.tzinfo]
        if config["timezone"] is not None:
            try:
                self.timezone = pytz.timezone(config["timezone"])
            except pytz.UnknownTimeZoneError as err:
                raise ConfigError("unknown timezone: " + str(err))
        elif self.utc:
            self.timezone = datetime.timezone.utc

        self.failsWhen = config.pop("failsWhen")
        self.onFailure = config.pop("onFailure")
        self.onPermanentFailure = config.pop("onPermanentFailure")
        self.onSuccess = config.pop("onSuccess")

        self.env_file = config.pop("env_file")
        self.environment = config.pop("environment")
        if self.env_file is not None:
            try:
                file_environs = parse_environment_file(self.env_file)
            except OSError as e:
                raise ConfigError("Could not load env_file: {}".format(e))
            else:
                # unpack variables in dictionaries
                config_environs = {
                    env["key"]: env["value"] for env in self.environment
                }
                # update file values with config ones
                file_environs.update(config_environs)
                # replace environment
                self.environment = [
                    {"key": key, "value": value}
                    for key, value in file_environs.items()
                ]

        self.executionTimeout = config.pop("executionTimeout")
        self.killTimeout = config.pop("killTimeout")
        self.statsd = config.pop("statsd")

        self.uid = None
        self.gid = None

        user = config.pop("user", None)
        if user is not None:
            if isinstance(user, int):
                self.uid = user
            else:
                try:
                    pw = getpwnam(user)
                    self.uid = pw.pw_uid
                    self.gid = pw.pw_gid
                except KeyError:
                    raise ConfigError("User not found: {!r}".format(user))

        group = config.pop("group", None)
        if group is not None:
            if isinstance(group, int):
                self.gid = group
            else:
                try:
                    self.gid = getgrnam(group).gr_gid
                except KeyError:
                    raise ConfigError("Group not found: {!r}".format(group))

        if self.uid is not None or self.gid is not None:
            if os.geteuid() != 0:
                raise ConfigError(
                    "Job {} wants to change user or group, "
                    "but yacron is not running as superuser".format(self.name)
                )


def parse_environment_file(path: str) -> Dict[str, str]:
    """
    Parse environment variables from file.

    Handles comments (lines starting with ``#``) and blank lines.
    Variables must be specified in ``VARIABLE_NAME=CONTENT`` format.

    :param path: Path to the environment file.
    :raise ConfigError: If a line in the file is not parsable
        (the ``=`` key-value separation character is missing).
    :raise OSError: If an error occurred while opening the file at ``path``.
    :return: key-value map of environment variables.
    """
    environ: Dict[str, str] = {}

    with open(path, "r") as env_file:
        # file parsing
        # you may want to use the `dotenv` library to do the job
        for line in env_file.readlines():
            line = line.strip(" ").rstrip("\n")
            if line.startswith("#") or not line:
                continue
            if "=" not in line:
                raise ConfigError(
                    "Invalid line in env_file: '{}'".format(line)
                )
            key, value = line.split("=", 1)
            key = key.strip(" ")
            value = value.strip(" ")
            environ[key] = value

    return environ


@dataclass
class YacronConfig:
    jobs: List[JobConfig]
    web_config: Optional[WebConfig]
    job_defaults: JobDefaults
    logging_config: Optional[LoggingConfig]


def parse_config_string(data: str, path: str) -> YacronConfig:
    try:
        doc = strictyaml.load(data, CONFIG_SCHEMA, label=path).data
    except YAMLError as ex:
        raise ConfigError(str(ex))

    inc_defaults_merged: dict = {}
    jobs = []
    webconf = WebConfig(doc["web"]) if "web" in doc else None
    logging_conf = LoggingConfig(doc["logging"]) if "logging" in doc else None
    for include in doc.get("include", ()):
        inc_path = os.path.join(os.path.dirname(path), include)
        inc_config = parse_config_file(inc_path)
        inc_defaults_merged = dict(
            mergedicts(inc_defaults_merged, inc_config.job_defaults)
        )
        jobs.extend(inc_config.jobs)
        if inc_config.web_config:
            if webconf:
                raise ConfigError("multiple web configs")
            webconf = inc_config.web_config
        if inc_config.logging_config:
            if logging_conf:
                raise ConfigError("multiple logging configs")
            logging_conf = inc_config.logging_config
    defaults = dict(mergedicts(DEFAULT_CONFIG, inc_defaults_merged))
    defaults = dict(mergedicts(defaults, doc.get("defaults", {})))
    for config_job in doc.get("jobs", []):
        job_dict = dict(mergedicts(defaults, config_job))
        jobs.append(JobConfig(job_dict))
    return YacronConfig(
        jobs=jobs,
        web_config=webconf,
        job_defaults=JobDefaults(defaults),
        logging_config=logging_conf,
    )


def parse_config_file(
    path: str,
) -> YacronConfig:
    with open(path, "rt", encoding="utf-8") as stream:
        data = stream.read()
    return parse_config_string(data, path)


def parse_config(config_arg: str) -> YacronConfig:
    jobs = []
    config_errors = {}
    web_config = None
    web_config_source_fname = None
    if os.path.isdir(config_arg):
        for direntry in os.scandir(config_arg):
            base, ext = os.path.splitext(direntry.name)
            if base[0] in {"_", "."}:
                continue
            if ext in {".yml", ".yaml"}:
                try:
                    config = parse_config_file(direntry.path)
                except ConfigError as err:
                    config_errors[direntry.path] = str(err)
                except OSError as ex:
                    config_errors[config_arg] = str(ex)
                else:
                    jobs.extend(config.jobs)
                    if config.web_config is not None:
                        if web_config is None:
                            web_config = config.web_config
                            web_config_source_fname = direntry.path
                        else:
                            raise ConfigError(
                                "Multiple 'web' configurations found: "
                                "first in {}, now in {}".format(
                                    web_config_source_fname, direntry.path
                                )
                            )
    else:
        try:
            config = parse_config_file(config_arg)
        except OSError as ex:
            config_errors[config_arg] = str(ex)
        else:
            jobs.extend(config.jobs)
    if config_errors:
        raise ConfigError("\n---".join(config_errors.values()))
    return YacronConfig(
        jobs=jobs,
        web_config=config.web_config,
        job_defaults=config.job_defaults,
        logging_config=config.logging_config,
    )
